### ISSUE TITLE
Circular reference in TangoDevice

### ISSUE THREAD

**Issue by cmft** [taurus-org/taurus#952]
There is a circular reference when you call `addListener` method  of a TangoDevice.
It can be reproduce with this snippet.

```
import taurus
f = taurus.Factory()

def l(*args,**kw): print(args, kw)

print(111, f.tango_devs.keys())
d = taurus.Device('sys/tg_test/1')
d.addListener(l)
print(222, f.tango_devs.keys())
d = None
print(333, f.tango_devs.keys())
```

**Comment by cmft** [comment]
The problem is not related to `addListener` is due to create an Attribute  `self.stateObj`

I simplify the snippet:
```
import taurus
f = taurus.Factory()

print(f.tango_devs.keys(), f.tango_attrs.keys())
d = taurus.Device('sys/tg_test/1')
d.getStateObj()
print(f.tango_devs.keys(), f.tango_attrs.keys())
d = None
print(f.tango_devs.keys(), f.tango_attrs.keys())
```

**Comment by cpascual** [comment]
There is indeed a circualr reference, but I think it cannot be easily avoided without breaking some compatibility.

The problem is that the `stateObj` attribute needs to be referenced (not weakref) if `device.addListener`is called, because otherwise the event subscriptions will be cancelled when the stateObj is  removed.

**Comment by cpascual** [comment]
IMHO, all this requires a major refactoring to be solved.

One possible approach would be to replace the internal implementation of the device event subscription (which currently uses `.stateObj` which is a  `taurus.core.tango.TangoAttribute`) and use the `taurus.core.tango.TangoDevice.deviceObject ` instead (which is a `tango.DeviceProxy`). Then `.stateObj`does not need to keep a cache

**Comment by cpascual** [comment]
I do not think that it is advisable to do such a refactoring in a hurry just before the Jul20 release... so I am moving it to the next milestone
